### PR TITLE
Enabling profile support -  Added separate modules for flink version 1.17 

### PR DESCRIPTION
**Pull request by prashastia**
To enable flink profiles, all common utilities were moved to `flink-connector-bigquery-common`. For this the following measures were taken
1. Flink dependency `flink-core` and `flink-annotations` has been added to make sure annotations and certain utilities like `SerializableSupplier`, `BiConsumerWithException`, `Preconditions`, `FunctionWithException` is supported.
2. GoogleCredentialSupplier
* Uses [BaseEncoding]([http://org.apache.flink.shaded.curator5.com.google.common.io.BaseEncoding)
* Replaced with [Base64.getDecoder()](https://docs.oracle.com/javase/8/docs/api/java/util/Base64.html).
* Another tentative alternative could have been [BaseEncoding](https://github.com/google/guava/blob/master/guava/src/com/google/common/io/BaseEncoding.java) however, flink does not allow `common.io` methods to coexist and prefers usage of `flink-shaded-guava` methods which would cause dependency on flink, so replaced with `java.utils.Base64`
* Both (`BaseEncoding` and `Base64`) support [RFC-4648](https://www.ietf.org/rfc/rfc4648.txt) format encoding.
3. BigQueryUtils and BigQueryServicesImpl require flink version logging for which `FlinkVersion` and `EnvironmentInformation` have been used. This is replaced by adding a similar implementation as that of `EnvironmentInformation`.
* To obtain the flink version, we use `EnvironmentInformation.getVersion()`. Following method chain is followed:
    * `getVersion()` -> `getVersionsInstance().projectVersion`
    * `getVersionsInstance()` -> `VersionsHolder.INSTANCE`
    * `VersionsHolder` -> `new Versions()`
* `Versions()` has the following implementation:
    ```java
    private static final String PROP_FILE = ".flink-runtime.version.properties";
    ClassLoader classLoader = EnvironmentInformation.class.getClassLoader();
    try (InputStream propFile = classLoader.getResourceAsStream(PROP_FILE)){
        if (propFile != null){
            Properties properties = new Properties();
            properties.load(propFile);
            projectVersion = getProperty(properties, "project.version", UNKNOWN);
        }
    }
    
    private String getProperty(Properties properties, String key, String defaultValue){
        String value = properties.getProperty(key);
        if (value == null || value.charAt(0) == '$'){
            return defaultValue;
        }
        return value;
    }
    ```
* A very similar implementation has been added under `FlinkVersion.getVersion()`.
    ```java
    public static String getVersion() {
        ClassLoader classLoader = FlinkVersion.class.getClassLoader();
        try (InputStream propFile = classLoader.getResourceAsStream(PROP_FILE)){
            Properties properties = new Properties();
            properties.load(propFile);
            String projectVersion = properties.getProperty("project.version");
            if (projectVersion == null || projectVersion.charAt(0) == '$') {
                return FlinkVersion.UNKNOWN;
            }
            return projectVersion;
        } 
        catch (IOException e){
            LOG.error(String.format("Could not obtain Flink Version.%nError: %s", e.getMessage()));
        }
        return null;
    }
    ```
4. Reordering of the file structure:
File structure has been moved and modified to be:

[flink-bigquery-connector](https://github.com/prashastia/flink-bigquery-connector)
– [flink-connector-bigquery-common](https://github.com/prashastia/flink-bigquery-connector/tree/maven-profiles-setup/flink-connector-bigquery-common)
– [flink-1.17-connector-bigquery](https://github.com/prashastia/flink-bigquery-connector/tree/maven-profiles-setup/flink-connector-bigquery-flink-1.17)
– – [flink-connector-bigquery](https://github.com/prashastia/flink-bigquery-connector/tree/maven-profiles-setup/flink-connector-bigquery-common)
– – [flink-connector-bigquery-example](https://github.com/prashastia/flink-bigquery-connector/tree/maven-profiles-setup/flink-connector-bigquery-examples/flink-connector-bigquery-example-flink-1.17)
– – [flink-connector-bigquery-integration-test](https://github.com/prashastia/flink-bigquery-connector/tree/maven-profiles-setup/flink-connector-bigquery-integration-test/flink-connector-bigquery-integration-test-flink-1.17)

/gcbrun

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -365,6 +279,16 @@ under the License.
                 <!-- activate API compatibility checks -->
                 <groupId>io.github.zentol.japicmp</groupId>
                 <artifactId>japicmp-maven-plugin</artifactId>
+				<!--TODO: Remove this -->
+				<configuration>
+					<dependencies>
+						<dependency>
+							<groupId>org.apache.flink</groupId>
+							<artifactId>flink-annotations</artifactId>
```
Comment (prashastia) [inline pom.xml:287]: japicmp: Pluguin and its requirements.

File: flink-connector-bigquery-common/src/main/java/com/google/cloud/flink/bigquery/common/utils/GoogleCredentialsSupplier.java [hunk 2]
```diff
@@ -16,20 +16,21 @@
...
+import com.google.cloud.flink.bigquery.common.utils.flink.annotations.Internal;
 
 import java.io.ByteArrayInputStream;
 import java.io.FileInputStream;
 import java.io.IOException;
 import java.io.UncheckedIOException;
+import java.util.Base64;
 import java.util.Optional;
 
+// TODO: Get this verified java.util.Base64 instead of
```
Comment (prashastia) [inline flink-connector-bigquery-common/src/main/java/com/google/cloud/flink/bigquery/common/utils/GoogleCredentialsSupplier.java:31]: This needs to be verified.

File: cloudbuild/presubmit/presubmit.sh [hunk 3]
```diff
@@ -29,13 +29,13 @@ cd /workspace
...
   init)
-    $MVN clean install -DskipTests
+    $MVN clean install -DskipTests -Pflink_1.17
     exit
     ;;
 
   # Run unit & integration tests
   tests)
-    $MVN clean clover:setup verify clover:aggregate clover:clover -Pclover
+    $MVN clean clover:setup verify clover:aggregate clover:check clover:clover -Pflink_1.17,clover
```
Comment (prashastia) [inline cloudbuild/presubmit/presubmit.sh:38]: clover:check goal enables checking of the coverage percentage specified with the expected coverage.

File: flink-connector-bigquery-flink-1.17/pom.xml [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>com.google.cloud.flink</groupId>
+        <artifactId>flink-connector-bigquery-parent</artifactId>
+        <version>${revision}</version>
+    </parent>
+
+    <artifactId>flink-connector-bigquery-flink-1.17</artifactId>
```
Comment (jayehwhyehentee) [inline flink-connector-bigquery-flink-1.17/pom.xml:30]: Pls rename to `flink-1.17-connector-bigquery-parent`

File: flink-connector-bigquery-flink-1.17/pom.xml
Comment (jayehwhyehentee) [inline flink-connector-bigquery-flink-1.17/pom.xml:1]: Pls rename folder to `flink-1.17-connector-bigquery`

Comment (jayehwhyehentee) [reply flink-connector-bigquery-flink-1.17/pom.xml:1]: Similar format throughout

File: flink-connector-bigquery-flink-1.17/flink-connector-bigquery/pom.xml [hunk 5]
```diff
@@ -23,155 +23,116 @@ under the License.
 
     <parent>
         <groupId>com.google.cloud.flink</groupId>
-        <artifactId>flink-connector-bigquery-parent</artifactId>
-        <version>0.1.0-preview</version>
+        <artifactId>flink-connector-bigquery-flink-1.17</artifactId>
```
Comment (jayehwhyehentee) [inline flink-connector-bigquery-flink-1.17/flink-connector-bigquery/pom.xml:26]: Rename to `flink-1.17-connector-bigquery`

File: pom.xml [hunk 6]
```diff
@@ -66,35 +62,19 @@ under the License.
...
-            <groupId>org.apache.flink</groupId>
-            <artifactId>flink-shaded-force-shading</artifactId>
-        </dependency>
-
         <!-- Root dependencies for all projects -->
         <dependency>
             <groupId>dev.failsafe</groupId>
             <artifactId>failsafe</artifactId>
         </dependency>
+        <dependency>
```
Comment (yerramalli) [inline pom.xml:74]: As this is used only in 1.17, move this dependency to the 1.17 module.

File: flink-connector-bigquery-common/src/main/java/com/google/cloud/flink/bigquery/common/utils/flink/annotations/Internal.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+import java.lang.annotation.Target;
+
+/**
+ * Annotation to mark methods within stable, public APIs as an internal developer API.
+ *
+ * <p>Developer APIs are stable but internal to Flink and might change across releases.
+ */
+@Documented
+@Target({ElementType.TYPE, ElementType.METHOD, ElementType.CONSTRUCTOR, ElementType.FIELD})
+@Public
```
Comment (yerramalli) [inline flink-connector-bigquery-common/src/main/java/com/google/cloud/flink/bigquery/common/utils/flink/annotations/Internal.java:32]: This is a copy of the source from Apache Flink lib, as this is not being extened it is not the correct approach to simply copy the interface or annotation. Instead, remove this completely and directly add the dependency to the flink lib.

**Review by yerramalli — CHANGES_REQUESTED** [review]

File: flink-connector-bigquery-common/src/main/java/com/google/cloud/flink/bigquery/common/utils/GoogleCredentialsSupplier.java [hunk 8]
```diff
@@ -66,8 +65,18 @@ private static Credentials createCredentialsFromAccessToken(String accessToken)
...
+            // But that would cause dependency on flink, so replaced with java.utils.Base64
+            // Both support RFC-4648 (https://www.ietf.org/rfc/rfc4648.txt)
+            // Links:
+            // 1. BaseEncoding:
+            // https://guava.dev/releases/17.0/api/docs/com/google/common/io/BaseEncoding.html#base64()
+            // 2. Base64:
+            // https://docs.oracle.com/javase/8/docs/api/java/util/Base64.html
             return GoogleCredentials.fromStream(
-                    new ByteArrayInputStream(BaseEncoding.base64().decode(key)));
+                    new ByteArrayInputStream(Base64.getDecoder().decode(key)));
```
Comment (prashastia) [inline flink-connector-bigquery-common/src/main/java/com/google/cloud/flink/bigquery/common/utils/GoogleCredentialsSupplier.java:79]: @yerramalli Please have a look at this replacement.

File: flink-connector-bigquery-common/src/main/java/com/google/cloud/flink/bigquery/services/BigQueryUtils.java [hunk 9]
```diff
@@ -80,7 +80,7 @@ public static Bigquery.Builder newBigqueryBuilder(CredentialsOptions options) {
                         new HttpCredentialsAdapter(options.getCredentials()))
                 .setApplicationName(
                         "BigQuery Connector for Apache Flink version "
-                                + EnvironmentInformation.getVersion());
+                                + FlinkVersion.current().toString());
```
Comment (prashastia) [inline flink-connector-bigquery-common/src/main/java/com/google/cloud/flink/bigquery/services/BigQueryUtils.java:83]: Please have a look at this replacement as well.

File: flink-connector-bigquery-common/src/main/java/com/google/cloud/flink/bigquery/common/utils/flink/annotations/Internal.java (hunk 7 above)
Comment (prashastia) [reply flink-connector-bigquery-common/src/main/java/com/google/cloud/flink/bigquery/common/utils/flink/annotations/Internal.java]: ACK, Fixed.

File: pom.xml (hunk 6 above)
Comment (prashastia) [reply pom.xml]: ACK, Fixed.

**Review by jayehwhyehentee — APPROVED** [review]

File: flink-connector-bigquery-common/src/main/java/com/google/cloud/flink/bigquery/common/utils/GoogleCredentialsSupplier.java (hunk 8 above)
Comment (yerramalli) [reply flink-connector-bigquery-common/src/main/java/com/google/cloud/flink/bigquery/common/utils/GoogleCredentialsSupplier.java:79]: Looks good, but please add unit tests to verify different types of encrypted keys.

**Review by yerramalli — APPROVED** [review]

**Pull request merged**